packages/types/index.d.ts

```diff
@@ -0,0 +1,610 @@
+import { Capabilities } from '@wdio/types';
```

Comment:
is this file hand-rolled or auto-generated?

Reply:
it's nice that we can reuse some of the work wdio has done, but internally we really don't care. internally it should just be:

```ts
{[key: string]: any}
```

maybe i've missed it and it's there somewhere already.

what if we add a new capability that wdio doesn't know about yet? what if a third party driver author adds a capability that wdio will never add to their list?

Reply:
Then I'd say we should provide somewhere to define those things that do not occur in wdio's capabilities.

But in general, more specificity is better.  Any _optional_ properties should be typed, because they won't prevent someone from providing the wrong type, otherwise.  Further, you get inline documentation on the properties, where provided.

So I guess I'm in disagreement about using `any` here.  Even if we did avoid reusing wdio's types, there's still stuff that is absolutely required that both consumers and contributors should be aware of (e.g. W3C caps have a certain shape, `platformName` is required, etc.).  It may be that at some point in the control flow we don't really care any more, but it matters for public APIs.

I think we can reuse the W3C caps in a way which gives us more flexibility.  First, we can redefine what, e.g., `AppiumCapabiltiies` are by doing something like:

```ts
// untested
import type {AppiumCapabilities as BaseAppiumCapabilities} from '@wdio/types/capabilities';

interface OtherCapabilities {
 // whatever we need to add
 'appium:foo': string
}

export type AppiumCapabilities = BaseAppiumCapabilities & OtherCapabilities;
```

We can also use a generic to allow drivers to optionally specify their own capabilities:

```ts
export interface Driver<C extends AppiumCapabilities = AppiumCapabilities> {
  validateDesiredCaps(caps: C): boolean;
  // etc
}
```

An implementing class could then do something like:

```ts
interface MyCaps extends AppiumCapabilities {
  'stuff': 'things'
}

class MyMainClassDriver extends BaseDriver implements Driver<MyCaps> {

}
```

Of course, minding the types is completely optional.

What do you think?

Reply:
Maybe we should put some effort into defining our own capability types then. I'm just not confident that WDIO's current ones are correct or will remain so even if they are.

Another issue to consider is that when we deal with capabilities internally, they don't have `appium:` in front, because this is stripped out in capabilities processing, but when users send capabilities to an Appium server (in which case they probably shouldn't be interfacing with these type definitions because they'll be using a client anyway, like WDIO, and consuming that client's type interface instead of ours), they do.

If we want to take this on, we'll need to hunt through the code to find the following types of capabilities mentioned so we can do the right thing with them:

- capabilities used by BaseDriver
- capabilities used by AppiumDriver
- capabilities used by a platform driver (which should be moved to a type definition in the platform driver instead)

And yes, in our internal code we'll always need to assume that the capabilities object could have other random things that the user has sent in. There are probably very few capabilities that we can definitely assume will always be present (I think platformName, automationName might be the only ones we can universally assume).

Reply:
I've come up with something I think will be workable here.  `Capabilities` for internal use, and `W3CCapabilities` for external use.  We build these types off of wdio's types, but they are very loose, and we are free to extend them as we wish.  This provides autocompletion and _only_ errors if the type is actually _wrong_ (e.g., a `number` when a `string` was expected).  `W3CCapabilities` must correspond (roughly) to the standard.